active/ItemsSourceView-Indexof/ItemsSourceView-IndexOf.md

```diff
@@ -0,0 +1,63 @@
+in this spec. -->
+The Xaml [ItemsRepeater](https://docs.microsoft.com/uwp/api/Microsoft.UI.Xaml.Controls.ItemsRepeater) control displays a list of objects using a specified template (similar to a [ListView](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.ListView)). You provide the list of objects with the [ItemsRepeater.ItemsSource](https://docs.microsoft.com/uwp/api/Microsoft.UI.Xaml.Controls.ItemsRepeater.ItemsSource) property. The list need only be an IIterable but can be an IVector for better perf. ItemsRepeater exposes a read-only [ItemsRepeater.ItemsSourceView](https://docs.microsoft.com/uwp/api/Microsoft.UI.Xaml.Controls.ItemsRepeater.ItemsSourceView) property that provides a view on this list (an [ItemsSourceView](https://docs.microsoft.com/uwp/api/Microsoft.UI.Xaml.Controls.ItemsSourceView)), which has helpers like Count and GetAt over what might be just an IIterable.
+
+Currently the ItemsSourceView class does not expose an IndexOf function. However during development, this method was used multiple times internally on derivatives of the class. To make the usage of this existing function easier on subclasses, this spec adds an IndexOf function to the ItemsSourceView.
+
+# Description
+
+The IndexOf function returns the index of the given element in the items source the ItemsSourceView is used for.
+Note that if an element is not contained in the items source, the function will return -1.
+> Spec note: returning -1 matches the behavior of for example IndexOf on [UIElementCollection](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.UIElementCollection) and List<T>.
+# Examples
+C# example
+
+```c#
+/// ************** Behavior ************ ///
+
+var items = new List<string>() { "One", "Two" };
+var itemsSourceView = new ItemsSourceView(items);
+var secondItem = itemsSourceView.GetAt(1);
+Assert.AreEqual(1, itemsSourceView.IndexOf(secondItem);
+
+
+/// ****** Use case in event handler ****** ///
+
+public void ItemInvoked(object sender, RoutedEventArgs e)
+{
+  int index = myItemsSourceView.IndexOf(sender); 
+
+  // Do something with index, here e.g. :
+  Debug.WriteLine( $"Item {index} of items source was invoked");
+}
+
+```
+
+
+# Remarks
+
+If an element is not present in the collection, the method returns -1.
+
+# API Notes
+ItemsSourceView.IndexOf(IInspectable item): Returns index in the items source. If an item is not in the items source, it will return -1. Null is an acceptable value for IndexOf and calling IndexOf with null will return the index of null in the ItemsSource or -1 if the ItemsSource does not contain null.
+
+If the ItemsSourceView's underlying collection supports [IVector<T>.IndexOf](https://docs.microsoft.com/uwp/api/Windows.Foundation.Collections.IVector-1.IndexOf) or [IVectorView<T>.IndexOf](https://docs.microsoft.com/uwp/api/Windows.Foundation.Collections.IVectorView-1.IndexOf), it will be used and that value returned.
```

Comment:
These interfaces are specifically IInspectable, not T, I assume?

Reply:
Apparently they are not, however calling them still works as expected (at least in our unit tests).

Reply:
It looks like in the implementation it's an IVector<IInspectable>. If the unit test is passing a List<T>, that class implements IList, and so the collection will project to WinRT as implementing both IVector<IInspectable> and IVector<T>. Do we actually use IVectorView's IndexOf, or does that fall back to the IIterable?

Reply:
The order of casting is: IVector->IBindableVector->IIterable->IBindableIterable
The first cast that succeeds will be used. So actually we would ignore any pure IVectorView. An C# IList would be used as IVector then.

Reply:
And specifically IVector<IInspectable>?

Reply:
If you pass an IVector, we use that IVector directly. According to the docmentation, the .Net projection of IVector<T> is IList, so IList would be used as IVector.

Reply:
The collection projections are complicated because in c# we want to use the language's standard interfaces. There isn't actually an IVector without the T, and the c# List&lt;T&gt; _class_ makes things even more interesting by implementing both IList&lt;T&gt; and IList.

| WinRT | C# |
| - | - |
| IVector&lt;T&gt; | IList&lt;T&gt; |
| IVectorView&lt;T&gt; | IReadOnlyCollection&lt;T&gt; |
| IBindableVector | IList |

I think diving into all of that is getting too complicated for this method description. It sounds like the correct technical description is that it uses the underlying IndexOf if the collection supports IVector&lt;T&gt; or IBindableVector or IVector&lt;IInspectable&gt;, otherwise it enumerates in order using IIterable&lt;IInspectable&gt; or IBindableIterable. And then translate this into c# terms, and I think it's just getting too complicated.

How about we just leave this at:

> If the ItemsSourceView's underlying collection supports a compatible IndexOf, it will be used and that value returned. Otherwise the collection will be enumerated and the index of the first matching item returned.

And then if looks too vague in the docs we can add a link on "compatible" to a page that generally describes the supported collection interfaces and how they project to different languages.

Reply:
Yes that whole projection situation is to much for the method description, I agree,

However, your proposed text is not entirely correct, since we still use Vector.IndexOf, if there is no suitable IndexOf we could use. So I think either we should change the way of handling this or we should change the section to something more like this:

> If the ItemsSourceView's underlying collection supports a compatible IndexOf, it will be used and that value returned. Otherwise the first matching value from its enumeration will be returned and IndexOf uses object identity. Note that for strings and value types, object identity is performed on the boxed value.

What do you think about this @mikehillberg ?

Reply:
OK, looks good to me :)

Reply:
Great, doc's updated now.